promql/functions.go

```diff
@@ -130,6 +134,209 @@ func extrapolatedRate(vals []parser.Value, args parser.Expressions, enh *EvalNod
+
+	// Cannot compute rate with 0 or 1 data points.
+	if len(points) < 2 {
+		return enh.Out
+	}
+
+	scrapeIntervalMsec := inferScrapeInterval(points) //** milliseconds
+
+	firstPoint := 0
+	var lastValue float64
+
+	// The 0 Fix: Check for fresh pod start case where we have no point before the range,
+	// meaning the caller failed to find any before the range in the lookback interval.
+	// Treat this as if we had a 0 to start.
+	firstPointToRangeStartMsec := rangeStartMsec - points[0].T
+	if firstPointToRangeStartMsec < 0 {
+		lastValue = 0.0 // include points[0].V in resultValue
+	} else {
+		// If the point before the range is too far from rangeStart, drop it.
+		if firstPointToRangeStartMsec > scrapeIntervalMsec { //** There's 0 slop here, so we could drop the point even if it was scraped 1 msec early
+			firstPoint = 1
+		}
+		lastValue = points[firstPoint].V
+	}
+
+	// Cannot compute rate with 0 or 1 data points.
+	if len(points)-firstPoint < 2 {
+		return enh.Out
+	}
+
+	// Simplest result: the difference beteween first and last values.
+	resultValue := points[len(points)-1].V - lastValue
+
+	// Debug logging
+	log.Printf("extendedRate0: isRate: %t, firstPoint: %d, scrapeIntervalMsec: %.1f", isRate, firstPoint, float64(scrapeIntervalMsec)/1000.0)
+	log.Printf("extendedRate0: enh range: %.3f...%.3f (offset %.3f); starting result: %.1f; lastValue: %.1f",
+		float64(rangeStartMsec)/1000.0, float64(rangeEndMsec)/1000.0, float64(vs.Offset)/1000.0, resultValue, lastValue)
+
+	{
+		buffer := new(bytes.Buffer)
+		for i, point := range points {
+			if i == elideSamplesAfter && len(points)-1 > elideSamplesAfter {
+				fmt.Fprintf(buffer, "...")
+			} else if i > elideSamplesAfter && len(points)-i > elideSamplesAfter {
+				continue
+			} else {
+				if i > 0 {
+					fmt.Fprintf(buffer, ", ")
+				}
+				fmt.Fprintf(buffer, "[%.3f,%.1f]", float64(point.T)/1000.0, point.V)
+			}
+		}
+		log.Println("extendedRate0: samples:", buffer.String())
+	}
+
+	// Correct for any case where a process restart caused the increase to turn negative.
+	for _, point := range points[firstPoint:] {
+		step := point.V - lastValue
+		if step < 0 {
+			resultValue += -step
+			log.Printf("extendedRate0: correcting for step: %.1f", step)
+		}
+		lastValue = point.V
+	}
+
+	// If the first or last sampled point is outside the range boundaries +/-, assume we are missing a sample at
+	// that edge and extrapolate from the sampled range up to the requested range.
+	{
+		sampledStartMsec := points[firstPoint].T
+		sampledEndMsec := points[len(points)-1].T
+		if int64(math.Abs(float64(rangeStartMsec-sampledStartMsec))) < scrapeIntervalMsec ||
+			firstPointToRangeStartMsec < 0 {
+			sampledStartMsec = rangeStartMsec // Close enough, so snap to start
+		}
+		if int64(math.Abs(float64(rangeEndMsec-sampledEndMsec))) < scrapeIntervalMsec {
+			sampledEndMsec = rangeEndMsec // Close enough, so snap to end
+		}
+		if sampledStartMsec != rangeStartMsec || sampledEndMsec != rangeEndMsec {
```

Comment:
The above is trying to make this extrapolation pretty rare, because when it happens, it often winds up exaggerating the slope because it sees an Integer step over a shorter timeframe.